File: 0056-merge-intervals/0056-merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    intervals.sort((a, b) => a[0] - b[0]);
    const ans = [];
    let currentInterval = intervals[0];

    for(let i = 1; i<intervals.length; i++) {
        if(currentInterval[1] >= intervals[i][0]) {
            currentInterval[0] = Math.min(currentInterval[0], intervals[i][0]);
            currentInterval[1] = Math.max(currentInterval[1], intervals[i][1]);
        }
        else {
            ans.push(currentInterval);
            currentInterval = intervals[i];
        }
    }
    ans.push(currentInterval);
    
    return ans;
};

File: 0056-merge-intervals/0056-merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        int idx = 0;

        for(int i = 1; i<intervals.size(); i++) {
            if(intervals[idx][1] >=  intervals[i][0]) {
               intervals[idx][0] = min(intervals[idx][0], intervals[i][0]);
               intervals[idx][1] = max(intervals[idx][1], intervals[i][1]);
            }
            else {
                idx++;
                intervals[idx][0] = intervals[i][0];
                intervals[idx][1] = intervals[i][1];
            }
        }

        intervals.resize(idx + 1);
        return intervals;
    }
};